add document for bucktization

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38119 add document for bucktization**

This is for (#37435).
Demo is here:
https://glaringlee.github.io/generated/torch.searchsorted.html
https://glaringlee.github.io/generated/torch.bucketize.html

Differential Revision: [D21517392](https://our.internmc.facebook.com/intern/diff/D21517392)

glaringlee committed:
add document for bucktization

[ghstack-poisoned]

glaringlee committed:
Update on "add document for bucktization"




[ghstack-poisoned]

zou3519 commented on torch/_torch_docs.py:
```diff
@@ -7754,3 +7754,82 @@ def merge_dicts(*dicts):
+bucketize(input, boundaries, out_int32=False, right=False, out=None) -> Tensor
+
+Returns a new tensor, same size as :attr:`input`, containing the result of applying
+either lower_bound or upper_bound (depends on :attr:`right`) to each value in :attr:`input`
```
Are "lower_bound" / "upper_bound" formal ways to describe this behavior? I'm not familiar with this terminology and it took me a while to understand exactly what was being conveyed here; if it is regular terminology then I'm fine with keeping it.

The way numpy describes it is clearer to me: https://docs.scipy.org/doc/numpy/reference/generated/numpy.digitize.html

"Return the indices of the [buckets] to which each value in input belongs."

glaringlee replied:
i was trying to follow https://www.tensorflow.org/api_docs/python/tf/searchsorted
Will take a look at the numpy one,

glaringlee replied:
updated

zou3519 replied:
I feel like some C++ user wrote the tf documentation lol. It reads much better now, thank you

zou3519 commented on torch/_torch_docs.py:
```diff
@@ -7754,3 +7754,82 @@ def merge_dicts(*dicts):
+           r"""
+searchsorted(sorted_sequence, values, out_int32=False, right=False, out=None) -> Tensor
+
+Returns a new tensor, same size as :attr:`values`, containing the result of applying
```
Ditto here, the numpy version of this function https://docs.scipy.org/doc/numpy/reference/generated/numpy.searchsorted.html 's documentation reads better to me

glaringlee replied:
same as above, will take a look at

glaringlee replied:
updated

zou3519 commented on torch/_torch_docs.py:
```diff
@@ -7754,3 +7754,82 @@ def merge_dicts(*dicts):
+    out (Tensor, optional): the output tensor, must be the same size as :attr:`values` if provided.
+
+.. note:: If your use case is always 1-D sorted sequence, :func:`torch.bucketize` is preferred,
+          it has less internal check, slightly better performance. 
```
nit: "If your use case is always 1-D sorted sequence, :func:`torch.bucketize` is preferred because it has slightly better performance" -> we don't really have to tell users about the internal details

glaringlee replied:
I think we should give user some level of the internal details on why we recommend one api over another, this will make people more comfortable to switch. Numpy doc did this as well.
https://docs.scipy.org/doc/numpy/reference/generated/numpy.digitize.html ->
"Note that as the order of the arguments are reversed, the side must be too. The searchsorted call is marginally faster, as it does not do any monotonicity checks."

glaringlee committed:
Update on "add document for bucktization"


This is for (#37435).
Demo is here:
https://glaringlee.github.io/generated/torch.searchsorted.html
https://glaringlee.github.io/generated/torch.bucketize.html



[ghstack-poisoned]

glaringlee committed:
Update on "add document for bucktization"


This is for (#37435).
Demo is here:
https://glaringlee.github.io/generated/torch.searchsorted.html
https://glaringlee.github.io/generated/torch.bucketize.html



[ghstack-poisoned]

zou3519 commented on torch/_torch_docs.py:
```diff
@@ -7754,3 +7754,126 @@ def merge_dicts(*dicts):
+
+Find the indices from the sorted 1-D tensor :attr:`boundaries` such that, if the corresponding values 
+in :attr:`input` were inserted before the indices, the order of :attr:`boundaries` would be preserved. 
+Return a new tensor, same size as :attr:`input`. The returned index satisfies the following rules:
```
nit: "Return a new tensor, same size as :attr:`input`"  reads better as "Return a new tensor with the same size as :attr:`input`" (the comma in the middle is weird to look at).

glaringlee replied:
done

zou3519 commented on torch/_torch_docs.py:
```diff
@@ -7754,3 +7754,126 @@ def merge_dicts(*dicts):
+bucketize(input, boundaries, out_int32=False, right=False, out=None) -> Tensor
+
+Find the indices from the sorted 1-D tensor :attr:`boundaries` such that, if the corresponding values 
+in :attr:`input` were inserted before the indices, the order of :attr:`boundaries` would be preserved. 
```
nit: I still prefer the bin terminology, but the language used in the current PR does state what the function does.

For the bin terminology, I'd write the following: "Returns the indices of the buckets to which each value in the input tensor belongs, where the boundaries of the buckets are set by `boundaries`".

glaringlee replied:
updated

zou3519 commented on torch/_torch_docs.py:
```diff
@@ -7754,3 +7754,126 @@ def merge_dicts(*dicts):
+
+Find the indices from the sorted 1-D tensor :attr:`boundaries` such that, if the corresponding values 
+in :attr:`input` were inserted before the indices, the order of :attr:`boundaries` would be preserved. 
+Return a new tensor, same size as :attr:`input`. The returned index satisfies the following rules:
```
I would also explain what the "right=False" does in words (the math below makes it clear, but it takes a while to parse). In words: "If right is False (default), then the right boundary is closed. More formally, the returned index satisfies the following rules:"

glaringlee replied:
done, I think it should be 'left boundary is closed'

zou3519 commented on torch/_torch_docs.py:
```diff
@@ -7754,3 +7754,126 @@ def merge_dicts(*dicts):
+    input (Tensor or Scalar): N-D tensor or a Scalar containing the search value(s).
+    boundaries (Tensor): 1-D tensor, must contain a monotonically increasing sequence.
+    out_int32 (bool, optional): indicate the output data type. torch.int32 if True, torch.int64 otherwise. 
+                                Default value is False, ie. default output data type is torch.int64.
```
nit: "i.e." instead of "ie."

zou3519 commented on torch/_torch_docs.py:
```diff
@@ -7754,3 +7754,126 @@ def merge_dicts(*dicts):
+    right (bool, optional): if False, return the first suitable location that is found. If True, return the 
+                            last such index. If no suitable index found, return 0 for non-numerical value 
+                            (eg. nan, inf) or the size of :attr:`boundaries` (one pass the last index). 
+                            In other words, if False, getting lower bound index 
```
nit: "if False, gets the lower bound index" (instead of "getting"). Ditto for the next sentence

zou3519 commented on torch/_torch_docs.py:
```diff
@@ -7754,3 +7754,126 @@ def merge_dicts(*dicts):
+           r"""
+searchsorted(sorted_sequence, values, out_int32=False, right=False, out=None) -> Tensor
+
+Find the indices from the *innermost* dimension of :attr:`sorted_sequence` such that, if the 
```
Funnily enough, here I like the sorting terminology better than the bin terminology. I guess it depends on the context (what the function is named)

zou3519 commented on torch/_torch_docs.py:
```diff
@@ -7754,3 +7754,126 @@ def merge_dicts(*dicts):
+Find the indices from the *innermost* dimension of :attr:`sorted_sequence` such that, if the 
+corresponding values in :attr:`values` were inserted before the indices, the order of the 
+corresponding *innermost* dimension within :attr:`sorted_sequence` would be preserved. 
+Return a new tensor, same size as :attr:`input`. The returned index satisfies the following rules: 
```
nit: "Returns a new tensor with the same size as :attr:`input`". The comma in the middle of the sentence is weird

zou3519 commented on torch/_torch_docs.py:
```diff
@@ -7754,3 +7754,126 @@ def merge_dicts(*dicts):
+    out (Tensor, optional): the output tensor, must be the same size as :attr:`values` if provided.
+
+.. note:: If your use case is always 1-D sorted sequence, :func:`torch.bucketize` is preferred, 
+          because it has less dimension check, slightly better performance.
```
nit: "because it has less dimension check, slightly better performance" -> "because it has fewer dimension checks resulting in slightly better performance"

zou3519 reviewed (approved):
I had some comments to make the text read better. They're non-blocking but please read. Other than that, LGTM

glaringlee commented:
@zou3519 Thanks a lot, will adopt your comments in, much better

glaringlee committed:
Update on "add document for bucktization"


This is for (#37435).
Demo is here:
https://glaringlee.github.io/generated/torch.searchsorted.html
https://glaringlee.github.io/generated/torch.bucketize.html



[ghstack-poisoned]
